File: Array/TwoPointers/productExceptSelf.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        int cnt_zero = 0;
        for (auto &x: nums) if (x==0) cnt_zero ++;
        
        if (cnt_zero == 1) {
            int p = 1, ind_zero;
            vector<int> res(n, 0);
            for (int i = 0; i < n; ++i) {
                if (nums[i] != 0) p *= nums[i];
                else ind_zero = i;
            }
            res[ind_zero] = p;
            return res;
        }
        else if (cnt_zero > 1) return vector<int> (n, 0);
        
        vector<int> res (n, 1);
        vector<int> left_prod (n,1);
        vector<int> right_prod (n,1);

        for (int i = 1; i < n; ++i) {
            left_prod[i] = left_prod[i-1]*nums[i-1];       
        }
        
        for (int i = n-2; i >= 0; --i) {
            right_prod[i] = right_prod[i+1]*nums[i+1];
        }
        for (int i = 0; i < n; ++i) res[i] = left_prod[i] * right_prod[i];
        
        return res;
    }
};

File: Array/TwoPointers/productExceptSelf.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        
        n = len(nums)
        if nums.count(0) == 1:
            p = 1
            res = [0 for i in range(n)]
            for i in range(n):
                if nums[i] != 0:
                    p = p * nums[i]
                else:
                    id = i
            res[id] = p
            return res
        elif nums.count(0) > 1:
            return [0 for i in range(n)]
            
        res = [1 for i in range(n)]
        l = [1 for i in range(n)]
        r = [1 for i in range(n)]
        for i in range(1, n):
            l[i] = l[i-1] * nums[i-1]
        for i in range(n-2, -1, -1):
            r[i] = r[i+1] * nums[i+1]
            # print(l)
            # print(r)
        for i in range(n):
            res[i] = l[i] * r[i]
        return res